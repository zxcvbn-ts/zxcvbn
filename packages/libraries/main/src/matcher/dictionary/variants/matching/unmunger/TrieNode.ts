export default class TrieNode {
  constructor(public parents: string[] = []) {}

  // eslint-disable-next-line no-use-before-define
  children: Map<string, TrieNode> = new Map()

  subs?: string[]

  addSub(key: string, ...subs: string[]): TrieNode {
    const firstChar = key.charAt(0)
    if (!this.children.has(firstChar)) {
      this.children.set(firstChar, new TrieNode([...this.parents, firstChar]))
    }
    let cur = this.children.get(firstChar)!
    for (let i = 1; i < key.length; i += 1) {
      const c = key.charAt(i)
      if (!cur.hasChild(c)) {
        cur.addChild(c)
      }
      cur = cur.getChild(c)!
    }
    cur.subs = (cur.subs || []).concat(subs)
    return this
  }

  getChild(child: string): TrieNode | undefined {
    return this.children.get(child)
  }

  isTerminal(): boolean {
    return !!this.subs
  }

  addChild(child: string): void {
    if (!this.hasChild(child)) {
      this.children.set(child, new TrieNode([...this.parents, child]))
    }
  }

  hasChild(child: string): boolean {
    return this.children.has(child)
  }
}
