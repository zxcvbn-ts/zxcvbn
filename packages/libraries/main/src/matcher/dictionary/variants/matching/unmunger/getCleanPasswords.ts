import TrieNode from './TrieNode'

interface GetAllSubCombosHelperOptions {
  substr: string
  buffer: string[]
  limit: number
  trieRoot: TrieNode
}

export interface PasswordChanges {
  letter: string
  substitution: string
}

export type IndexedPasswordChanges = PasswordChanges & {i: number}

export interface PasswordWithSubs {
  password: string
  changes: IndexedPasswordChanges[]
}

const getAllSubCombosHelper = ({
  substr,
  buffer,
  limit,
  trieRoot,
}: GetAllSubCombosHelperOptions): PasswordWithSubs[] => {
  const finalPasswords: PasswordWithSubs[] = []

  // eslint-disable-next-line max-statements
  const helper = (
    onlyFullSub: boolean,
    isFullSub: boolean,
    index: number,
    subIndex: number,
    changes: IndexedPasswordChanges[],
  ): void => {
    if (finalPasswords.length >= limit) {
      return
    }

    if (index === substr.length) {
      if (onlyFullSub === isFullSub) {
        finalPasswords.push({ password: buffer.join(''), changes })
      }
      return
    }

    // first, exhaust all possible substitutions at this index
    const nodes: TrieNode[] = []
    let cur = trieRoot
    for (let i = index; i < substr.length; i += 1) {
      const character = substr.charAt(i)
      cur = cur.getChild(character)!
      if (!cur) {
        break
      }
      nodes.push(cur)
    }
    let hasSubs = false
    // iterate backward to get wider substitutions first
    for (let i = index + nodes.length - 1; i >= index; i -= 1) {
      cur = nodes[i - index]
      if (cur.isTerminal()) {
        hasSubs = true
        const subs = cur.subs!
        // eslint-disable-next-line no-restricted-syntax
        for (const sub of subs) {
          buffer.push(sub)
          const newSubs = changes.concat({
            i: subIndex,
            letter: sub,
            substitution: cur.parents.join(''),
          })

          // recursively build the rest of the string
          helper(onlyFullSub, isFullSub, i + 1, subIndex + sub.length, newSubs)
          // backtrack by ignoring the added postfix
          buffer.pop()
          if (finalPasswords.length >= limit) {
            return
          }
        }
      }
    }
    // next, generate all combos without doing a substitution at this index
    // if a partial substitution is requested or there are no substitutions at this index
    if (!onlyFullSub || !hasSubs) {
      const firstChar = substr.charAt(index)
      buffer.push(firstChar)
      helper(onlyFullSub, isFullSub && !hasSubs, index + 1, subIndex + 1, changes)
      buffer.pop()
    }
  }

  // only full substitution
  helper(true, true, 0, 0, [])
  // only partial substitution
  helper(false, true, 0, 0, [])

  return finalPasswords
}

const getCleanPasswords = (
  string: string,
  limit: number,
  trieRoot: TrieNode,
): PasswordWithSubs[] => {
  return getAllSubCombosHelper({
    substr: string,
    buffer: [],
    limit,
    trieRoot,
  })
}
export default getCleanPasswords
