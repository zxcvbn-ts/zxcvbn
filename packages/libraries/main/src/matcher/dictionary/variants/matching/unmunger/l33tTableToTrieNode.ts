import { OptionsL33tTable } from '../../../../../types'
import TrieNode from './TrieNode'

export default (l33tTable: OptionsL33tTable, triNode: TrieNode) => {
  Object.entries(l33tTable).forEach(([letter, substitutions]) => {
    substitutions.forEach((substitution) => {
      triNode.addSub(substitution, letter)
    })
  })
  return triNode
}
