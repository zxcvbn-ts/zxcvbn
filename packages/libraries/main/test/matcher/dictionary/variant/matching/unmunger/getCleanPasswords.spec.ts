import l33tTableToTrieNode from '../../../../../../src/matcher/dictionary/variants/matching/unmunger/l33tTableToTrieNode'
import TrieNode from '../../../../../../src/matcher/dictionary/variants/matching/unmunger/TrieNode'
import getCleanPasswords from '../../../../../../src/matcher/dictionary/variants/matching/unmunger/getCleanPasswords'

const trieNode = l33tTableToTrieNode(
  {
    a: ['@', '4'],
    u: ['|_|'],
    m: ['^^', 'nn', '2n', '/\\\\/\\\\'],
  },
  new TrieNode(),
)
describe('getCleanPasswords', () => {
  it('should get correct clean passwords', () => {
    expect(getCleanPasswords('P4|_|$nn4rd', 100, trieNode)).toMatchSnapshot()
  })

  it('should limit the substitutions correctly', () => {
    expect(getCleanPasswords('P4|_|$nn4rd', 2, trieNode)).toMatchSnapshot()
  })
})
