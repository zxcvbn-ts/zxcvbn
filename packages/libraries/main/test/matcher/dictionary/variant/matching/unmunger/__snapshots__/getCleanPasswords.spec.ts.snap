// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCleanPasswords should get correct clean passwords 1`] = `
[
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
      {
        "i": 4,
        "letter": "m",
        "substitution": "nn",
      },
      {
        "i": 5,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "Pau$mard",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
      {
        "i": 4,
        "letter": "m",
        "substitution": "nn",
      },
    ],
    "password": "Pau$m4rd",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
      {
        "i": 6,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "Pau$nnard",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
    ],
    "password": "Pau$nn4rd",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 6,
        "letter": "m",
        "substitution": "nn",
      },
      {
        "i": 7,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "Pa|_|$mard",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 6,
        "letter": "m",
        "substitution": "nn",
      },
    ],
    "password": "Pa|_|$m4rd",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
      {
        "i": 8,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "Pa|_|$nnard",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "Pa|_|$nn4rd",
  },
  {
    "changes": [
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
      {
        "i": 4,
        "letter": "m",
        "substitution": "nn",
      },
      {
        "i": 5,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "P4u$mard",
  },
  {
    "changes": [
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
      {
        "i": 4,
        "letter": "m",
        "substitution": "nn",
      },
    ],
    "password": "P4u$m4rd",
  },
  {
    "changes": [
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
      {
        "i": 6,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "P4u$nnard",
  },
  {
    "changes": [
      {
        "i": 2,
        "letter": "u",
        "substitution": "|_|",
      },
    ],
    "password": "P4u$nn4rd",
  },
  {
    "changes": [
      {
        "i": 6,
        "letter": "m",
        "substitution": "nn",
      },
      {
        "i": 7,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "P4|_|$mard",
  },
  {
    "changes": [
      {
        "i": 6,
        "letter": "m",
        "substitution": "nn",
      },
    ],
    "password": "P4|_|$m4rd",
  },
  {
    "changes": [
      {
        "i": 8,
        "letter": "a",
        "substitution": "4",
      },
    ],
    "password": "P4|_|$nnard",
  },
  {
    "changes": [],
    "password": "P4|_|$nn4rd",
  },
]
`;

exports[`getCleanPasswords should limit the substitutions correctly 1`] = `
[
  {
    "changes": [
      {
        "i": 0,
        "letter": "w",
        "substitution": "vv",
      },
    ],
    "password": "w",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "f",
        "substitution": "v",
      },
      {
        "i": 1,
        "letter": "f",
        "substitution": "v",
      },
    ],
    "password": "ff",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "f",
        "substitution": "v",
      },
      {
        "i": 1,
        "letter": "u",
        "substitution": "v",
      },
    ],
    "password": "fu",
  },
]
`;

exports[`getCleanPasswords should substitute to multiple symbols correctly 1`] = `
[
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "@",
      },
      {
        "i": 4,
        "letter": "fi",
        "substitution": "ﬁ",
      },
    ],
    "password": "pacific",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "a",
        "substitution": "@",
      },
    ],
    "password": "paciﬁc",
  },
  {
    "changes": [
      {
        "i": 4,
        "letter": "fi",
        "substitution": "ﬁ",
      },
    ],
    "password": "p@cific",
  },
  {
    "changes": [],
    "password": "p@ciﬁc",
  },
]
`;

exports[`getCleanPasswords should substitute to multiple variants correctly 1`] = `
[
  {
    "changes": [
      {
        "i": 0,
        "letter": "w",
        "substitution": "vv",
      },
    ],
    "password": "w",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "f",
        "substitution": "v",
      },
      {
        "i": 1,
        "letter": "f",
        "substitution": "v",
      },
    ],
    "password": "ff",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "f",
        "substitution": "v",
      },
      {
        "i": 1,
        "letter": "u",
        "substitution": "v",
      },
    ],
    "password": "fu",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "u",
        "substitution": "v",
      },
      {
        "i": 1,
        "letter": "f",
        "substitution": "v",
      },
    ],
    "password": "uf",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "u",
        "substitution": "v",
      },
      {
        "i": 1,
        "letter": "u",
        "substitution": "v",
      },
    ],
    "password": "uu",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "f",
        "substitution": "v",
      },
    ],
    "password": "fv",
  },
  {
    "changes": [
      {
        "i": 0,
        "letter": "u",
        "substitution": "v",
      },
    ],
    "password": "uv",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "f",
        "substitution": "v",
      },
    ],
    "password": "vf",
  },
  {
    "changes": [
      {
        "i": 1,
        "letter": "u",
        "substitution": "v",
      },
    ],
    "password": "vu",
  },
  {
    "changes": [],
    "password": "vv",
  },
]
`;
